### DISCUSSION TITLE
Keeping track of vertex IDs, is property_maps the right way? Surface Mesh

### DISCUSSION THREAD

**Discussion by MartensCedric** [CGAL/cgal#7727] [Q&A]
I have some code like this

```
    std::vector<Point_3> points(curve_points.begin(), curve_points.end());
    Polyhedron_3 poly;
    CGAL::convex_hull_3(points.begin(), points.end(), poly);
    Surface_mesh sm;
    CGAL::convex_hull_3(points.begin(), points.end(), sm);

    std::ofstream outfile(argv[2]);
    if (!outfile.is_open()) {
        std::cerr << "Failed to open output file." << std::endl;
        return 1;
    }

    /* // Write the vertices to the OBJ file */
    for (auto vertex : sm.vertices()) {
        Point_3 p = sm.point(vertex);
        outfile << "v " << p.x() << " " << p.y() << " " << p.z() << std::endl;
    }

    for (auto f : sm.faces()) {
        outfile << "f";
        for(auto vh : sm.vertices_around_face(sm.halfedge(f))) {
            outfile << " " << static_cast<size_t>(vh) + 1;
        }
        outfile << std::endl;
    }


    outfile << "l";

    for(int i = 1; i <= N; i++)
        outfile << " " << i;
    outfile.close();
```

If you notice carefully, I have added an `l` to my .obj output, which is a line. I have given some importance to the first N vertices.

However, I have noticed that the order I obtain the vertices from sm.vertices() is different than the one I have specified at the beginning (the original vector of points).  How can I keep track of vertex indices? I have found the page on property maps and I believe they might be useful to store the index of the points (See section 3: https://doc.cgal.org/latest/Point_set_3/index.html#Point_set_3_Usage). 

```
    Index_map index_map;
    Point_set point_set;
    bool success = false;
    std::tie(index_map, success) = point_set.add_property_map<unsigned int>("index");
    assert(success);
    
    // 
    point_set.reserve(points.size());
    for(int i = 0; i < points.size(); i++)
    {
        Point_set::iterator it = point_set.insert(points[i]);
        index_map[*it] = i;
    }
```
However, I have had some difficulties on actually retrieving the index:

```
    for (int i = 0; i < points.size(); i++) {
        Point_3 p = points[i];
        outfile << "v " << p.x() << " " << p.y() << " " << p.z() << std::endl;
    }

    for (auto f : sm.faces()) {
        outfile << "f";
        for(auto vh : sm.vertices_around_face(sm.halfedge(f))) {
            Point_3 vertex = sm.point(vh);
            unsigned int old_index = index_map[vertex]; // doesn't compile
            outfile << " " << old_index + 1;
        }
        outfile << std::endl;
    }
```

I am not terribly surprised that it fails since the point set might need to hash Point_3s 

```
no match for ‘operator[]’ (operand types are ‘Index_map’ {aka ‘CGAL::Point_set_3<CGAL::Point_3<CGAL::Epick> >::Property_map<unsigned int>’} and ‘Point_3’ {aka ‘CGAL::Point_3<CGAL::Epick>’}
```

I found on this thread (https://stackoverflow.com/q/70582185/4048657) a comment mentioning: typedef CGAL::Polyhedron_3<Kernel,CGAL::Polyhedron_items_with_id_3>  Polyhedron;

Which seems a bit similar to what I want but I think I need it for CGAL::Surface_mesh<Point_3>, and CGAL::Surface_mesh_items_with_id_3 doesnt seem to exist.

So I'm a bit confused on how I'm supposed to keep track of my vertex IDs

**Comment by sloriot** [comment]
There is indeed no guarantee that points added to the mesh are added in the same order in the mesh.

I'm afraid there is no solution out of the box. You could generate a 3D triangulation like [here](https://doc.cgal.org/latest/Triangulation_3/Triangulation_3_2info_insert_with_transform_iterator_8cpp-example.html) to set the vertex indices corresponding to your input point.
Then using a modified version of [`convex_hull_3_to_face_graph()`](https://doc.cgal.org/latest/Convex_hull_3/group__PkgConvexHull3Functions.html#ga2750f7f197588ed643679835c748c671) you should be able to set the id back into a Surface mesh property map. See https://doc.cgal.org/latest/Surface_mesh/#title7
